Slack Notifications

chdecultot opened the pull request:
Hi,

This PR is a proposal to add Slack Notifications in ERPNext.
Using the Email Alerts feature, we could allow users to choose between sending an email or sending a Slack Notification, which is something a lot of companies would like to do. In the future, we could also add additional channels of course.

1. Create a Slack App
![new-slack-app](https://user-images.githubusercontent.com/4903591/40370612-f29660c0-5de0-11e8-85fd-10649fe2523f.png)

2. Create new Slack Webhook URLs
![slack-webhooks](https://user-images.githubusercontent.com/4903591/40370647-02569d22-5de1-11e8-895d-4d971002cb73.png)

3. Add the Slack Webhook URL to ERPNext
![slack-webhook-url](https://user-images.githubusercontent.com/4903591/40370671-0d673b7c-5de1-11e8-9d77-545e2b4780ba.png)

4. Configure your Email Alert to send the message to your Slack Channel based on document events
![slack-message](https://user-images.githubusercontent.com/4903591/40370710-25aa1592-5de1-11e8-9a7e-89e1879d36ef.png)



Maybe it could be interesting to rename Email Alerts to something more generic like Notification ? What do you think ?


The documentation is available with the Email Alert documentation in ERPNext: https://github.com/frappe/erpnext/pull/14184

chdecultot committed:
Slack notifications

chdecultot committed:
Slack notifications

chdecultot committed:
Codacy corrections

chdecultot committed:
Codacy corrections

codingCoffee committed:
fixed merge conflict

pratu16x7 committed:
Merge pull request #5591 from codingCoffee/merge_conflict

fixed merge conflict

pratu16x7 committed:
[query-perf-log] Add execution time for debug opt in frappe.db.sql (#5592)

* [query-perf-log] Add execution time for debug opt in frappe.db.sql

* Update database.py

Zlash65 committed:
[Enhance] New DocField - 'Fetch From' (#5563)

* add new docfield fetch_from - serves as add_fetch

* replace options.split with fetch_from.split

* copied data from options to fetch_from

* patch to update fetch_from for docfield, custom_field and property_setter

* update tests according to fetch_from change

sushantnadkar committed:
[fix] Button duplication on page trigger: refresh (#5564)

* [fix] custom button duplication on page trigger: refresh

* added data-label to buttons to filter them more easily

* added encodeURIComponent to data-label

* Update page.js

chdecultot committed:
Allow several Stripe accounts (#5573)

* Allow several Stripe accounts

* Addition of a sync patch

* Remove unused dependancy

gaurav-naik committed:
modified address and contact js (#4862) (#5586)

* modified aaddress and contact js

- temporary fix to handle route_history if docname is naming_series with a '/' in the name- permanant fix would be to modify route options
- PR linked to frappe/erpnext#12215

* considered cases with more '/'s in doc name

rohitwaghchaure committed:
[Fix] Hyperlink showing in the pdf print (#5590)

Zlash65 committed:
run docfield patch prior to other patches (#5594)

nabinhait committed:
Update activity_log.json

rmehta committed:
[minor] add pdfkit to requirements.txt

rmehta committed:
[fix] ellipsis in navbar brand #5599

rmehta committed:
[fix] calendar render for empty list, user_list.js

codingCoffee committed:
[fix] global-setup-help (#5603)

- custom jinja templates are taken if and only if a site is mentioned
- added test in travis as well

codingCoffee committed:
Prevents circular linking of docs (#5589)

* Prevents circular linking of docs

- also added test case to check if code to prevent circular linking of
doctypes works

* created a CircularLinkingError class

- this can be used later if circular linking error needs to be thrown
elsewhere

Zlash65 committed:
fix email query (#5605)

codingCoffee committed:
Merge branch 'hotfix'

codingCoffee committed:
bumped to version 10.1.33

codingCoffee committed:
Merge branch 'master' into develop

sharathc3 committed:
Jinja Customization from hooks.py bug fix (#5604)

* Jinja Customization from hooks.py bug fix

* Condition to check for site initialization added.

nabinhait committed:
fetch_from fixes

nabinhait committed:
fetch_from fixes in mapper

nabinhait committed:
Fetch From also applicable for Check fields (#5611)

Commit:
handle Int overflow (#5607)

* handle the int overflow

 - convert the int to the bigint if the length is more than 11

* proper formatting

schilgod committed:
Enhancement to allow User to login from any IP if two factor auth is enabled (#5209)

* Enhancement to allow login from any IP if two factor auth is enabled

* Resolve Conflicts

* optimize code

pratu16x7 committed:
[model][std-fields] make 'Created By' a User Link field (#5600)

Olawale1 commented:
Hi @chdecultot 

Awesome job!!!

Cheers!

rmehta committed:
[minor] strip special characters when making a fieldname

pratu16x7 committed:
[fix] cached timestamp in frappe.model.with_doctype() (#5613)

* [fix] cached timestamp in frappe.model.with_doctype()

* Update model.js

yimesgen committed:
Update country_info.json (#5616)

surajshetty3416 committed:
Treeview permissions (#5615)

* Replace frappe.db.sql to frappe.get_list to apply permissions (#5556)

* Replace frappe.db.sql with frappe.get_list
- Permission check for treeviews

* Optimize get_role_permissions
- Avoid setting of roles each time when is_perm_appicable is called

* Fix codacy
- Remove trailing whitespace

* Add parent filter

* Add ifnull check

codingCoffee committed:
[fix] contact to user creation patch (#5617)

- special charcters in name used to break user creation, this patch
prevents user creation in case of bad characters

rmehta committed:
[fix] for bench get-app

codingCoffee committed:
[fix] creates contact even in presence of bad char (#5618)

rmehta committed:
[fix] .travis.yml

saurabh6790 committed:
bumped to version 10.1.34

saurabh6790 committed:
Merge branch 'hotfix'

saurabh6790 committed:
[fix] resolved merge conflicts

rmehta committed:
[fix] bench_helper.py

netchampfaris committed:
fix: Don't add currency field if it does not exist (#5619)

closes #4995

Commit:
Update db_schema.py (#5622)

netchampfaris committed:
fix: get_contact_list in Email dialog (#5623)

nabinhait committed:
Fetch from property is now applicable for all fieldtypes except no_value_fields (#5624)

* Reload docperm

* Reload docperm

* Fetch from property is now applicable for all fieldtypes except no_value_fields

nabinhait committed:
changed patch order

nabinhait committed:
Patch fixes

nabinhait committed:
get root of nestedset docs

surajshetty3416 committed:
Workflow actions (#5598)

* Init workflow actions

- WIP code
- Add hooks to create and update
workflow action on document status changes.
- Tweak listview to show host document on list view item click
- Add new doctype "Pending Workflow Action"
- Add allow_self_approval check in Workflow master

* Rename doctypes

- Rename "Workflow Action" to "Workflow Action Master"
- Rename "Pending Workflow Action" to "Workflow Action"
- Remove irrelevant files

* Clear old workflow action on docstatus change

- Add check for user permission before creating a workflow action record
- Make doctype changes to support the Workflow Action  System

* Show host document only if workflow action status is open

* Send workflow action mail notifications

* Add patch to rename `tabWorkflow Action` to `tab Workflow Master`

- Rename table to retain data in `tabWorkflow Action` as it
will be overwritten with new structure on migration

* Enqueue email

* Remove unused files

* Add hook for custom has_permission

* Optimize send_workflow_action_email method

* Add permission_query_condition hook

* Show notification based on status

* Override get_form_link in list view

* Refactor workflow action

- Add action link
- update workflow action hooks

* Avoid repeated creation of Worklow Actions

- Update hooks entry to avoid unwanted method call

* Fix error with frappe.db.count usage

* Fix ui/ux for Workflow Action

- Fix action url and improve response on action button click
- Fix workflow action template style

* Fix Codacy

* Add self approval check

* Fix codacy

* Fix test

* Fix python version confilct and a permission error

* Fix incorrect 'this' reference

* Update with requested changes

* Add next action email template option

* Fix string formatting

* Refactor workflow Action

- Make process_workflow_action smaller

* Fix bugs and errors due to refactor

* Fix workflow field caching

* Add a workflow action test

* Fix bugs with email template

rmehta commented on frappe/email/doctype/email_alert/email_alert.json:
```diff
@@ -41,10 +42,111 @@
+   "label": "Slack Channel", 
+   "length": 0, 
+   "no_copy": 0, 
+   "options": "Slack Webhook URL", 
```
maybe we can configure any webhook for this?

rmehta commented on frappe/email/doctype/email_alert/email_alert.py:
```diff
@@ -171,13 +142,55 @@ def get_attachment(doc):
+				print_letterhead = ((attachments
+					and attachments[0].get('print_letterhead')) or False))
+
+		elif self.channel == 'Slack':
```
Make smaller functions. This function is too long.

chdecultot replied:
Sure I'll refactor it

rmehta commented on frappe/email/doctype/email_alert/email_alert.py:
```diff
@@ -171,13 +142,55 @@ def get_attachment(doc):
+					and attachments[0].get('print_letterhead')) or False))
+
+		elif self.channel == 'Slack':
+			slack_webhook_url = self.slack_webhook_url
```
should this be part of the Slack Setup?

chdecultot replied:
Sure I'll refactor it

chdecultot replied:
Actually the user needs to choose it's channel manually: one notification can be sent to one channel, the other one to another channel.

I could of course drop this variable which is not useful and write it like:

```
send_slack_message(
				webhook_url=self.slack_webhook_url,
				message=frappe.render_template(self.message, context),
				reference_doctype = doc.doctype,
				reference_name = doc.name
			)
```

rmehta reviewed (changes requested):
Should we also rename "Email Alert" to "Notification" ? I am sure SMS is next!

Alchez committed:
[Improvements] Auto-publish pages with date ranges (#5430)

* Add feature for auto-(un)publishing Web Pages

* Move some publish logic to frontend

* Cleanup some logic

* Remove arbitrary limits and add check for start date

tundebabzy committed:
DocType Naming With Multiple Fields (#5497)

* Update README.md

* refactor for better readability

adds ability to name by concatenating fields

add comments

PEP 8 fixes

throw error when field option autoname fails

adds concatenation option description to DocType doctype

* codacy

* remove explicit cast to `str`

* more pep8 fixes

* revert incorrect link

* code review - move internal functions to bottom of module

Commit:
[fix] linked with if there are multiple link field in the same doctype (#5587)

* [fix] linked with if there are multiple link field in the same doctype

* handle python 3 and fixed as per reviews

nabinhait committed:
Fixed sanitizing fields for tabDeleted Documents (#5329)

saurabh6790 committed:
[fix] Sanitize blacklisted functions

saurabh6790 committed:
Merge pull request #5633 from saurabh6790/sanitize_blacklisted_functions

[fix] Sanitize blacklisted functions

chdecultot commented:
Sure I'll rename Email Alert to Notification.
Should I keep it in the Email module of move it to another module (Core?) ?

chdecultot committed:
Rename Email Alert to Notification

chdecultot committed:
Merge issue correction

chdecultot commented:
Sorry issue with Git, I will open a new PR...
